### PR TITLE
[FLINK-19520][configuration] Add randomization of checkpoint config.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

With the larger refactoring of checkpoint alignment and the additional of more unaligned checkpoint settings, it becomes increasingly important to provide a large test coverage.

Unfortunately, adding sufficient test cases in a test matrix appears to be unrealistic: many of the encountered issues were subtle, sometimes caused by race conditions or unusual test configurations and often only visible in e2e tests.

Hence, we like to rely on all existing Flink tests to provide a sufficient coverage for checkpointing. However, as more and more options in unaligned checkpoint are going to be implemented in this and the upcoming release, running all Flink tests - especially e2e - in a test matrix is prohibitively expensive, even for nightly builds.

Thus, we want to introduce test randomization for all tests that do not use a specific checkpointing mode. In a similar way, we switched from aligned checkpoints by default in tests to unaligned checkpoint during the last release cycle.

To not burden the developers of other components too much, we set the following requirements:

- Randomization should be seeded in a way that both builds on Azure pipelines and local builds will result in the same settings to ease debugging and ensure reproducibility.
- Randomized options should be shown in the test log.
- Execution order of test cases will not influence the randomization.
- Randomization is hidden, no change on any test is needed.
- Randomization only happens during local/remote test execution. User deployments are not affected.
- Test developers are able to avoid randomization by explicitly providing checkpoint configs.

## Brief change log

- Add generic method to choose random values for checkpoint config.
- Use it for uninitialized "execution.checkpointing.unaligned".
- Adds flag to enable randomization in root pom. User code will not trigger randomization but chooses constant default values.

## Verifying this change

- Not covered (yet)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by rkhachatryan — COMMENTED** [review]
The randomization here only applies to tests that use `MiniClusterResource`, right?
I think this is a significant limitation.

Can we overcome it by reading some system property at the time of `ConfigOption` creation?
```
	public static final ConfigOption<Boolean> ENABLE_UNALIGNED =
		ConfigOptions.key("execution.checkpointing.unaligned")
			.booleanType()
			.defaultValue(System.getProperty("execution.checkpointing.unaligned", false)) // <-- here
```
The property can be set either 
- by maven/CI at the very beginning to a random value (seed="nanoTime")
- or by the developer or to a fixed value

I think it will be easier to reason about if it's set for the whole build (and logged in the beginning).

WDYT?

**Comment by AHeise** [comment]
> The randomization here only applies to tests that use `MiniClusterResource`, right?
> I think this is a significant limitation.
> 
> Can we overcome it by reading some system property at the time of `ConfigOption` creation?
> 
> ```
> 	public static final ConfigOption<Boolean> ENABLE_UNALIGNED =
> 		ConfigOptions.key("execution.checkpointing.unaligned")
> 			.booleanType()
> 			.defaultValue(System.getProperty("execution.checkpointing.unaligned", false)) // <-- here
> ```
> 
> The property can be set either
> 
>     * by maven/CI at the very beginning to a random value (seed="nanoTime")
> 
>     * or by the developer or to a fixed value
> 
> 
> I think it will be easier to reason about if it's set for the whole build (and logged in the beginning).
> 
> WDYT?

That was the very first approach but we didn't like that this means that all tests run only with that particular configuration, which has two issues:
- For a large enough test matrix, a certain configuration may happen very rarely. Assuming we have 10 different configuration options (aligned, unaligned with 0, 10s, 1m timeout x 0, 1kb, 1m size limit), then we have 10% chance for a particular combination to occur. But even when running 10 tests, you have 3.4% chance of a specific combination not occurring. If you add that to rarely occurring instabilities that are even now hard to detect, the overall goal of having a good coverage is probably not reached. It becomes worse if we have more interdependent value combinations related to checkpointing like different DSTL settings.
- Even for more commonly occurring issues that happen while refactoring/new feature, it's disadvantage to just have one configuration for all tests. You have to manually cycle through the relevant settings through setting the system property to "force" your luck. Here having more or less all combinations being executed by the same AZP run on your feature branch will probably already find you quite a few issues.

Btw (but this is an orthogonal discussion), I'd like to bind the randomization seed to commit id instead of timestamp, such that it's easy for us to debug into any issue (checkout particular commit and go). However, the current implementation might not be ideal in that regard (`EnvironmentalInformation` might not be updated without a full build).

**Comment by rkhachatryan** [comment]
Assuming that the priorities are:
1. increase probability of detecting an error
1. in particular, before merging - i.e. by committer either locally or remotely
1. improve the usability

The "MiniClusterResource" approach compared to the "properties" approach:
- is worse for (1) because a smaller subset of tests is randomized
- is better for (2) as you mention above
- is worse for (3) because it's not easy to see what else tests fail with a certain configuration (when I already know that some fail)

It also seems more complex though I haven't thought through how to set the property globally.

Additionally, I can imagine a "mandatory" set of combinations that must pass (in several runs) in order to release or be sure to merge.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (pnowojski) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:57]: So for the given commit id and given test name, the random selection choices will be always the same, on each run on any environment/machine?

If so, maybe put that (^^^) as a java doc for this `create` method?

**Review by pnowojski — COMMENTED** [review]
I think this change looks good to me.

Have you manually verified that the unaligned checkpoints are indeed randomly selected within a sing azure run? I mean not only checking what was logged in
```
LOG.info("Randomly selected {} for {}", value, option.key());
```
but via for example adding a temporary log message somewhere in the `StreamTask` or network stack what alignment option was used after all? I mean there could be a potential problem where for some (or all?) tests, the randomly selected value is still unintentionally overridden to some static value, yielding this change a no-op.

**Comment by pnowojski** [comment]
> Can we overcome it by reading some system property at the time of ConfigOption creation?

We didn't like this idea, as it would require to modify the production code.

**Comment by rkhachatryan** [comment]
> various drawbacks
Could you elaborate?

**Comment by pnowojski** [comment]
For one thing, that's a public api, so we wouldn't be able to easily change it. We could maybe mark it partially as `@Internal`, but I don't know how to do it in a clear way. Additionally it could just create confusion for the users, in the worst case they can start using it. Apart from that, if you need to modify production code for testing purposes, it usually means something is wrong with the design (having a testing mini cluster that randomly sets configuration options is a cleaner design). And also relaying on the `System.getProperty` is not the cleanest way.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:57]: Yes that was my intent: have it reproducible without changing anything by just checking out the particular commit that failed on AZP.

I'm double-checking that `EnvironmentInformation.getGitCommitId()` works as expected or try a different approach to inject commit id.

**Comment by AHeise** [comment]
> but via for example adding a temporary log message somewhere in the `StreamTask` or network stack what alignment option was used after all? I mean there could be a potential problem where for some (or all?) tests, the randomly selected value is still unintentionally overridden to some static value, yielding this change a no-op.

Good point, I have actually completely relied on the log message to check for the result. I'd repeat your suggestion.

Btw, it is my intention that an explicit value for the random options always take precedence. But I assume you want me to check for _unintentional_ overrides.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/util/TestLogger.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/util/TestLogger.java:71]:
```suggestion
	public final TestRule nameProvider = new TestNameProvider();
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:52]: This seems a bit...janky? Why can we not mutate the configuration within the MiniClusterResource?

Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:52]: I'm missing a hook where I can mutate the configuration per test method. `MiniClusterResource` is mostly used on class level. Any guidance is greatly appreciated.

Comment (zentol) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:52]: I see. I can't come up with a good alternative unfortunately.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java]: So I verified that randomized unaligned checkpoint configuration is actually picked up while creating the controllers.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java
Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java]: I moved it a bit outside into the factory, but that's the best that I came up with.

File: flink-runtime/pom.xml
Comment (zentol) [inline flink-runtime/pom.xml:512]: Why is it important for this to work in IntelliJ?

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:92]: hmm...like, we _could_ add a hook to setAsContext to pass the name/modify the config/pass a factory, but it'd still suck since the TestNameProvider must live within the `TestLogger`; we just can't get around that singleton access can we...

Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:92]: To make it independent of `TestLogger` we could use [RunListener](https://junit.org/junit4/javadoc/4.12/org/junit/runner/notification/RunListener.html) and then have a singleton listener to fetch the name.
However, I didn't go fully into this direction as it's quite a bit of fiddling in junit4 and I hope to eventually migrate to junit 5. Then I'd check if it can be done in a better way.

File: flink-runtime/pom.xml
Comment (AHeise) [reply flink-runtime/pom.xml:512]: I had mixed success when executing tests in IntelliJ. 
`EnvironmentInformation.getCommitId` depends on successful resource filtering of `.flink-runtime.version.properties` in `flink-runtime`, which worked for me fine with `mvn`.
However, in IntelliJ, quite often the `git.*` variables have not been replaced successfully. What's worse they have been unset even after having executed the task in `mvn resources -pl flink-runtime` successfully first.

I did not have a good explanation so I added some code to generate the `git.properties` first and then merge it into `.flink-runtime.version.properties`. I noticed that generating `git.properties` worked very reliably. However, something unexpected happened: `.flink-runtime.version.properties` was also now properly filtered. 🤷 

Since this is a bit hacky, an alternative could be to use two files to fill in the `EnvironmentInformation`: `git.properties` for git stuff and for `.flink-runtime.version.properties` for everything else.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java
Comment (zentol) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:92]: yeah I'd revisit it once junit5 is in.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:55]: What part of this class is not thread safe?

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:114]: braces! (I hope this will fail the CI, if not we might have to re-enable some checkstyle rules...)

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:45]: technically speaking these should be  `@VisibleForTesting`.

**Review by zentol — COMMENTED** [review]
approach-wise I think this looks good.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:55]: It doesn't adhere to the contract that it will produce the same randomized configuration for the same seeds if you start using it in parallel (the state is hidden in `Random` itself).

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java:45]: Technically you are correct - the best kind of correct. Fixed it.

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:59]:
```suggestion
    private static final Logger LOG = LoggerFactory.getLogger(PseudoRandomValueSelector.class);

    private final Function<Integer, Integer> randomValueSupplier;
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:61]:
```suggestion
    private static final long GLOBAL_SEED = (long) getGlobalSeed().hashCode() << 32;
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:78]:
```suggestion
        final long combinedSeed = GLOBAL_SEED | entryPointSeed.hashCode();
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:71]:
```suggestion
        final int choice = randomValueSupplier.apply(alternatives.length);
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java:103]:
```
Test randomization was enabled but test.randomization.seed was not configured, nor could the commit hash be retrieved from git or the EnvironmentInformation. Please set the test.randomization.seed property manually to make the build reproducible.
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:36]:
```suggestion
    private static final boolean RANDOMIZE_CHECKPOINTING_CONFIG =
```

File: flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java:93]:
```suggestion
        if (RANDOMIZE_CHECKPOINTING_CONFIG) {
```

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:57]:
```suggestion
        final PseudoRandomValueSelector valueSelector = PseudoRandomValueSelector.create("seed");
```

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:65]: assertThat(uniqueValues.size(); greaterThan(1))?

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:68]:
```suggestion
```

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:93]:
```suggestion
    /** Tests that the selector produces the same value for the same seed. */
```

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:102]: maybe add another set for the second value, to test that a sequence of selections is also deterministic.

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:114]: re-use `org.apache.flink.runtime.testutils.ZooKeeperTestUtils#runsOnCIInfrastructure` instead

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (AHeise) [reply flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:101]: I'm checking a triple of config options now.

File: flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java
Comment (zentol) [inline flink-test-utils-parent/flink-test-utils/src/test/java/org/apache/flink/streaming/util/PseudoRandomValueSelectorTest.java:49]: extend TestLogger

**Review by zentol — APPROVED** [review]

**Comment by AHeise** [comment]
AZP failure unrelated (https://issues.apache.org/jira/browse/FLINK-21103), merging.

**Pull request merged**
